docs/design.md

```diff
@@ -0,0 +1,508 @@
+
+### Preparing Ansible Contents
+
+To support Ansible run, Ansible provider requires the Ansible contents to be available before run. This may involve the work to retrieve Ansible contents from a remote place or to create a `playbook.yml` file including the inline content, then store into the local working directory.
+
+In Ansible provider, this is supported by implement the above logic in `Connect()`.
+Once an `AnsibleRun` resource is created, the reconciler will call the provider method `Connect()` to retrieve Ansible contents from the remote or generate inline playbook file which depends on how we define `AnsibleRun`.
+
+### Ansible Run Policy
+
+Once Ansible contents are available, we can start the Ansible run. Ansible provider supports a couple of run policies to fulfill different types of requirements. The policy is represented as annotation `ansible.crossplane.io/runPolicy`, which can be applied to `AnsibleRun` resource, to instruct the provider how to run the corresponding Ansible contents.
+
+#### Policy ObserveAndDelete
+
+This is the default policy and probably the most commonly used policy to manage Ansible run. When this policy is applied, the provider uses `Observe()` to handle the case when the managed resource `AnsibleRun` is present, and uses `Delete()` to handle the case when the resource is absent. Both `Observe()` and `Delete()` will call the same set of Ansible contents.
+
+![](images/ansible-run-policy-1.png)
+
+Here is an example to run an Ansible role using ObserveAndDelete policy:
+
+```yaml
+apiVersion: ansible.crossplane.io/v1alpha1
+kind: AnsibleRun
+metadata:
+  name: remote-example
+  annotation:
+    ansible.crossplane.io/runPolicy: ObserveAndDelete
+spec:
+  forProvider:
+    role: sample_namespace.sample_role
+  providerConfigRef:
+    name: provider-config-example
+```
+
+In order to differentiate the presence or absence of the `AnsibleRun` resource, a special variable is sent to Ansible when the Ansible contents start to run:
+
+```
+ansible_provider_meta.managed_resource.state = absence|presence
+```
+
+This variable belongs to a special set of variables maintained by Ansible provider where developers who create the Ansible contents can reference these variables when needed. For example, to run Ansible playbooks conditionally by checking the variable that indicates the presence or absence of `AnsibleRun` as below:
+
+```
+- include_tasks: setup-resource.yml
+  when: ansible_provider_meta.managed_resource.state == 'present'
+
+- include_tasks: cleanup-resource.yml
+  when: ansible_provider_meta.managed_resource.state == 'absent'
+```
+
+In future release, we should allow users to use arbitrary name for the variable that represents the presence or absence of the `AnsibleRun` resource, so that the Ansible contents maintained by user do not have to be coupled with or aware of Ansible provider.
+
```

Comment:
I think we need a more detailed scheme for this strategy :) 

I'm not sure this policy can emerge in the way crossplane operators are designed.
What I understood is the following:

1. **First point:**
 `each policy defined a new reconciliation strategy` good for Me  👍 

2. **Second point**
- the reality/desired state is handled by the  AnsibleRun spec 
- actual state is  AnsibleRun spec itself 

==> _actual state and reality are the same_!! ==> this can lead for multiple reconciling confusion

if I rephrase my words differently:

- This strategy answers the question : `should we execute the tasks or not ?` but to emerge into the crossplane ecosystem we should answer the following question : `is the task well executed or not?`

Reply:
Did not quite understand here as well. e.g.:

- What crossplane operators do you mean?
- What the difference among reality/actual/desired state?
- What does it mean: is the task well executed or not?

Could you elaborate a bit more?

Reply:
1)  crossplane operator = crossplane providers
2) What crossplane provider try to resolve is to ensure reconciliation for managed resource. 
* So the managed resource Spec ( e.g `Firewall` CR Spec in gcp provider) is the reality and the source of truth  
* The Firewall rule in IaaS Cloud  is the State of the managed Firewall managed resource.

So the gcp provider takes action to bring the managed resource actual state closer to the desired state.

if i go back to our ansible provider and try to map all of this to the reconciliation loop:
* the truth is the `AnsibleRun` CR because it handles **the strategy information** 
* The state of `AnsibleRun` CR (what the `observe()` watch )  is the existence of the CR itself 

=> the reality it is no longer role(s)/playbook(s) in the spec but the annotation 
=> the state is no longer the role/playbooks state but the existence of `AnsibleRun CR`

I understand that the lack of a state for the ansible compared to terraform push us to make  an intermediate step to determine the presence/absence of a resource but we should be careful how we have to emerge this in the crossplane ecosystem

Reply:
Now I understand what you are saying much better. Let me explain it and see if anything else I missed :-)

We do not drift from our goal. Compare w/ the gcp filrewall example, our `AnsibleRun` CR is equivalent to `Firewall` CR. Just as the `Spec` defines the source of truth for firewall, in `AnsibleRun`, mostly the `vars` defines the desired state or the source of truth. For example, below AnsibleRun relies on `my_namespace.openshift_cluster` to provision an OpenShift cluster and it defines how the cluster should look like in `vars`, e.g.: specifying the OpenShift version.

```yaml
apiVersion: ansible.crossplane.io/v1alpha1
kind: AnsibleRun
metadata:
  name: openshift-cluster
spec:
  forProvider:
    role: my_namespace.openshift_cluster
    vars:
      providerSecretRef: fyre-secret
      ocpVersion: "4.8.27"
      platform: "x"
      size: "large"
  providerConfigRef:
    name: provider-config-example
```

The annotation used to specify the strategy information is not part of the desired state or source of truth. It is really just a small chunk of metadata that instructs the Ansible provider how to trigger the Ansible role or playbook.

For the logic to handle the existence or absence of `AnsibleRun` CR, `Firewall` CR does it the same way. When the `Firewall` CR is deleted, the gcp provider should respond to delete the remote firewall resource on gcp. In our case, when the `AnsibleRun` CR is deleted, the Ansible provider will respond to call the Ansible role or playbook to delete whatever the remote resource.

Hope that clarifies the thing.

Reply:
Okey, this makes sense 👍 . Can you plz add these informations to the design. 
I think this will make the understanding of the ansible provider in the crossplane ecosystem much clearer

Reply:
@morningspace 
How will the Ansible provider user handle the addition of annotation(s)? are they mandatory? If yes What is the default policy? will it be added automatically by the controller

Reply:
> How will the Ansible provider user handle the addition of annotation(s)? are they mandatory?

No, they are not mandatory. The default policy would be the one that does not rely on check mode which will be auto-added by the controller if not specified. The reasons I am using annotation are that:

- To avoid the confusion for people who mix it w/ the desired state defined in `Spec`. It is just a small chunk of metadata that instructs the Ansible provider how to trigger the Ansible role or playbook.
- To avoid the overhead of API version upgrade in case we change the behavior later per user feedback. I think the idea of policy is still at early stage and may be changed after it is adopted by the community for a certain period of time. Instead of using annotation, if we add that into `Spec`, we will have to deal w/ API version upgrade to support backward compatibility or migration.